### PR TITLE
Examples on howto access XDP packet data at packet end

### PR DESCRIPTION
**Pull request by netoptimizer**
This is related to my work of changing BPF helper `bpf_xdp_adjust_tail` to allow tailgrow.

When doing this work I realized that writing BPF programs that access the last byte in the XDP packet is hard, as the verifier have some restrictions that can be challenging to workaround. This is also true when not using `bpf_xdp_adjust_tail`.

This PR contains two examples that works, but also two examples that demonstrate different ways the verifier fails and denies accessing the end-of-packets. I need someone to review, if my interpretation/explanation of why the verifier rejects these cases are true/correct.

### CODE REVIEW

**Comment by netoptimizer** [comment]
I'm going to merge this, to make it easier for others to git clone and test programs in experiment01-tailgrow.

This link should work after merging:
[experiment01-tailgrow](https://github.com/xdp-project/xdp-tutorial/tree/master/experiment01-tailgrow)

I still need review, especially for my comments in  [fail1](https://github.com/xdp-project/xdp-tutorial/blob/tailgrow02.public/experiment01-tailgrow/xdp_prog_fail1.c) and [fail2](https://github.com/xdp-project/xdp-tutorial/blob/tailgrow02.public/experiment01-tailgrow/xdp_prog_fail2.c)
(Master branch links: [fail1](https://github.com/xdp-project/xdp-tutorial/tree/master/experiment01-tailgrow/xdp_prog_fail1.c) and [fail2](https://github.com/xdp-project/xdp-tutorial/tree/master/experiment01-tailgrow/xdp_prog_fail2.c).)

And input if we can another BPF technique to access packet end.

**Comment by borkmann** [comment]
> I still need review, especially for my comments in [fail1](https://github.com/xdp-project/xdp-tutorial/blob/tailgrow02.public/experiment01-tailgrow/xdp_prog_fail1.c) and [fail2](https://github.com/xdp-project/xdp-tutorial/blob/tailgrow02.public/experiment01-tailgrow/xdp_prog_fail2.c)
> (Master branch links: [fail1](https://github.com/xdp-project/xdp-tutorial/tree/master/experiment01-tailgrow/xdp_prog_fail1.c) and [fail2](https://github.com/xdp-project/xdp-tutorial/tree/master/experiment01-tailgrow/xdp_prog_fail2.c).)

Could you try:
```
unsigned int offset = data_end - data;

offset &= 0x7fff;
data += offset;
if (data + 1 > data_end)
   goto out;
// 1 byte access to data[0]
```

File: experiment01-tailgrow/xdp_prog_fail1.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	if (pos + offset > data_end)
+		goto out;
+
+	/* Fails at this line with:
+	 *   "invalid access to packet, off=-1 size=1, R1(id=2,off=0,r=0)"
+	 *   "R1 offset is outside of the packet"
+	 *
+	 * Because verifer used offset==0 it thinks that we are trying
+	 * to access (data - 1), which is not within [data,data_end)
```
Comment (tohojo) [inline experiment01-tailgrow/xdp_prog_fail1.c:39]: If this is the case, shouldn't it help to check that offset > sizeof(*ptr)?

File: experiment01-tailgrow/xdp_prog_fail2.c [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	volatile void *pos;
+
+	pos = data_end;
+
+#pragma clang optimize off
+	if (pos - 1 > data_end)
+		goto out;
+#pragma clang optimize on
+
+	/* Verifier fails with: "pointer arithmetic on pkt_end prohibited"
```
Comment (tohojo) [inline experiment01-tailgrow/xdp_prog_fail2.c:27]: Any reason we couldn't allow pointer arithmetic on data_end as well? As long as all the other bounds checks are still enforced, of course...

File: experiment01-tailgrow/xdp_prog_kern.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+		/* Finding end of packet + offset, and bound access */
+		if ((void *)iphdr + ip_tot_len + offset > data_end) {
+			action = XDP_ABORTED;
+			goto out;
+		}
+
+		/* Point ts to end-of-packet, that have been offset extended */
+		ts = (void *)iphdr + ip_tot_len;
+		ts->magic = 0x5354; /* String "TS" in network-byte-order */
```
Comment (tohojo) [inline experiment01-tailgrow/xdp_prog_kern.c:88]: I think you're assuming a certain host byte ordering here, aren't you?

File: experiment01-tailgrow/xdp_prog_kern2.c [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+int _xdp_end_loop(struct xdp_md *ctx)
+{
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data     = (void *)(long)ctx->data;
+	unsigned char *ptr;
+	unsigned int i;
+	void *pos;
+
+	/* Assume minimum length to reduce loops needed a bit */
+	unsigned int offset = MIN_LEN;
```
Comment (tohojo) [inline experiment01-tailgrow/xdp_prog_kern2.c:28]: If you're going to optimise, couldn't you technically make the loop a binary search instead? :)

Comment (netoptimizer) [reply experiment01-tailgrow/xdp_prog_kern2.c:28]: Yes, to binary-search, but I wonder/worry how the BPF-verifiers bounded-loops verification would handle that ;-)

**Comment by tohojo** [comment]
Jesper Dangaard Brouer <notifications@github.com> writes:

> @netoptimizer commented on this pull request.
>
>
>
>> + * loop until reaching data_end.  This is needed as the verifier only allows
> + * accessing data via advancing the position of the data pointer. The bounded
> + * loop with a max number of iterations allows the verifier to see the bound.
> + */
> +
> +SEC("xdp_end_loop")
> +int _xdp_end_loop(struct xdp_md *ctx)
> +{
> +	void *data_end = (void *)(long)ctx->data_end;
> +	void *data     = (void *)(long)ctx->data;
> +	unsigned char *ptr;
> +	unsigned int i;
> +	void *pos;
> +
> +	/* Assume minimum length to reduce loops needed a bit */
> +	unsigned int offset = MIN_LEN;
>
> Yes, to binary-search, but I wonder/worry how the BPF-verifiers
> bounded-loops verification would handle that ;-)

As long as the maximum number of loop iterations is fixed that should be
fine, shouldn't it?

**Comment by netoptimizer** [comment]
> @borkmann says: Could you try:
> 
> ```
> unsigned int offset = data_end - data;
> 
> offset &= 0x7fff;
> data += offset;
> if (data + 1 > data_end)
>    goto out;
> // 1 byte access to data[0]
> ```

This code construct does pass the verifier, but the running code will **drop all packets**, because offset is the packet length, and the `+1` puts one byte past the end-of-packet. Thus, the if-check will drop all the packets (I tested this).

**Pull request merged**